gc/gc-base/src/test/java/org/projectnessie/gc/roundtrip/TestMarkAndSweep.java

```diff
@@ -0,0 +1,320 @@
+    final Instant now = Instant.now();
+    final Instant cutOff;
+    final Instant maxFileModificationTime;
+    final long newestToDeleteMillis;
+    final long tooNewMillis;
+
+    MarkAndSweep(long numCommits, int numKeysAtCutOff, long numExpired) {
+      this.numCommits = numCommits;
+      this.numKeysAtCutOff = numKeysAtCutOff;
+      this.numExpired = numExpired;
+      this.cutOff = now.minus(numCommits - numExpired, ChronoUnit.SECONDS);
+      this.maxFileModificationTime = now.minus(1, ChronoUnit.SECONDS);
+      this.newestToDeleteMillis = maxFileModificationTime.toEpochMilli();
+      this.tooNewMillis = maxFileModificationTime.toEpochMilli() + 1;
+    }
+
+    @Override
+    public String toString() {
+      return "CommitsAndContentsSimulation{"
+          + "numCommits="
+          + numCommits
+          + ", numKeysAtCutOff="
+          + numKeysAtCutOff
+          + ", numExpired="
+          + numExpired
+          + '}';
+    }
+
+    String numToHash(long l) {
+      return String.format("%016x", l);
+    }
+
+    String numToContentId(long l) {
+      return cidPrefix + l;
+    }
+
+    URI numToBaseLocation(long l) {
+      return basePath.resolve(l + "/");
+    }
+
+    long hashToNum(Reference ref) {
+      return Long.parseLong(ref.getHash(), 16);
+    }
+
+    long contentIdToNum(String cid) {
+      return Long.parseLong(cid.substring(cidPrefix.length()));
+    }
+
+    long baseLocationToNum(URI path) {
+      String p = path.getPath();
+      int l = p.lastIndexOf('/');
+      int l1 = p.lastIndexOf('/', l - 1);
+      return Long.parseLong(p.substring(l1 + 1, l));
+    }
+
+    Content numToContent(long l) {
+      return IcebergTable.of(
+          numToBaseLocation(l).resolve("metadata-" + l).toString(), l, 1, 2, 3, numToContentId(l));
+    }
+
+    ContentKey numToContentKey(long l) {
+      return ContentKey.of("content", Long.toString(l));
+    }
+
+    LogEntry numToLogEntry(long l) {
+      return LogEntry.builder()
+          .commitMeta(
+              CommitMeta.builder()
+                  .commitTime(now.minus(numCommits - l, ChronoUnit.SECONDS))
+                  .hash(numToHash(l))
+                  .message("hello")
+                  .build())
+          .parentCommitHash(numToHash(l - 1))
+          .addOperations(Put.of(numToContentKey(l), numToContent(l)))
+          .build();
+    }
+
+    long liveContentsCount() {
+      return numCommits + numKeysAtCutOff - numExpired;
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("markAndSweep")
+  public void markAndSweep(MarkAndSweep markAndSweep) throws Exception {
+
+    InMemoryPersistenceSpi storage = new InMemoryPersistenceSpi();
+    LiveContentSetsRepository repository =
+        LiveContentSetsRepository.builder().persistenceSpi(storage).build();
+    IdentifyLiveContents identify =
+        IdentifyLiveContents.builder()
+            .contentTypeFilter(
+                new ContentTypeFilter() {
```

Comment:
I hope we provide some default implementations for `ContentTypeFilter`, `contentToReference()` and `contentToFiles`() to the users, right now it looks like big code needed for calling Identify() and expire()

Reply:
_Users_ should really _not_ code directly against GC.

Reply:
As we saw in the mailing list, some people want to use API instead of CLI.